### PR TITLE
RPC v0 implementation

### PR DESCRIPTION
**Pull request by jlmcmchl**
As discussed on discord, this implements the RPC v0 spec in nt-network. Work is required in the main nt-rs project to make this useful.

### CODE REVIEW

**Comment by jlmcmchl** [comment]
Updated - finished server + client RPC implementations. 

Review @Redrield ?

File: src/proto/server.rs [hunk 1]
```diff
@@ -9,6 +10,9 @@ use nt_network::{
 };
 use std::collections::HashMap;
 use std::net::SocketAddr;
+use std::panic;
+use std::panic::RefUnwindSafe;
```
Comment (Redrield) [inline src/proto/server.rs:13]: `use std::panic::{self, RefUnwindSafe};`

File: src/proto/client/conn.rs [hunk 2]
```diff
@@ -262,6 +262,12 @@ fn handle_packet(packet: ReceivedPacket, state: &Arc<Mutex<ClientState>>) -> cra
                 state.lock().unwrap().clear_entries();
             }
         }
+        ReceivedPacket::RpcResponse(rpc) => {
+            let mut state = state.lock().unwrap();
+            if let Some(callback) = state.rpc_callbacks.remove(&rpc.unique_id) {
+                callback(rpc.result);
```
Comment (Redrield) [inline src/proto/client/conn.rs]: Evaluate the callback asynchronously, we don't want user code accidentally stalling the network loop

File: src/proto/client.rs [hunk 3]
```diff
@@ -24,6 +26,8 @@ pub struct ClientState {
     connection_callbacks: MultiMap<ConnectionCallbackType, Box<ConnectionAction>>,
     pub(crate) pending_entries: HashMap<String, Sender<u16>>,
     pub(crate) packet_tx: UnboundedSender<Box<dyn Packet>>,
+    rpc_callbacks: HashMap<u16, Box<RpcCallback>>,
+    next_id: u16,
```
Comment (Redrield) [inline src/proto/client.rs]: Specify the ID is for RPC

File: src/proto/server.rs [hunk 4]
```diff
@@ -56,6 +62,28 @@ impl ServerState {
...
+        let id = self.create_entry(data).try_next().unwrap().unwrap();
+        self.rpc_actions.insert(id, Box::new(callback));
+    }
+
+    pub fn call_rpc(&self, id: u16, parameter: Vec<u8>) -> Vec<u8> {
+        let entry = self.entries.get(&id).unwrap();
+        let rpc = self.rpc_actions.get(&id).unwrap();
+
+        match entry.value {
+            EntryValue::RpcDefinition(_) => match panic::catch_unwind(|| rpc(parameter)) {
```
Comment (Redrield) [inline src/proto/server.rs]: Evaluate the RPC body asynchronously, as with the client callback

File: src/proto/server/conn.rs [hunk 5]
```diff
@@ -263,6 +263,19 @@ where
                         }
                     }
                 }
+                ReceivedPacket::RpcExecute(rpc) => {
+                    let state = state.lock().unwrap();
+                    let result = state.call_rpc(rpc.entry_id, rpc.parameter.clone());
+
+                    state.clients.iter().for_each(|(_, tx)| {
+                        tx.unbounded_send(Box::new(RpcResponse::new(
```
Comment (Redrield) [inline src/proto/server/conn.rs]: RpcResponse should only be sent to the client that initiated the call

**Review by Redrield — CHANGES_REQUESTED** [review]

**Comment by Redrield** [comment]
@jlmcmchl See the changes I've requested

**Comment by jlmcmchl** [comment]
Hey, thanks for the feedback. I've implemented most of your changes, but I'm struggling to get the server-side async working. Here's what the code looks like currently:

    ReceivedPacket::RpcExecute(rpc) => {
        let state = state.lock().unwrap();
        let client = Box::pin(Box::new(state.clients.get(&addr).unwrap().clone()));

        match state.rpc_actions.get(&rpc.entry_id) {
            Some(func) => {
                let func = Box::pin(Box::new(func.clone()));
                thread::spawn(move || {
                    Pin::into_inner(client)
                        .unbounded_send(Box::new(RpcResponse::new(
                            rpc.entry_id,
                            rpc.unique_id,
                            func(rpc.parameter),
                        )))
                        .unwrap();
                });
            }
            None => {
                Pin::into_inner(client)
                    .unbounded_send(Box::new(RpcResponse::new(
                        rpc.entry_id,
                        rpc.unique_id,
                        Vec::new(),
                    )))
                    .unwrap();
            }
        }
    }

And the error:

        Checking nt v1.0.3 (C:\Dev\jlmcmchl\nt-rs)
    error[E0597]: `state` does not live long enough
      --> src\proto\server\conn.rs:269:33
        |
    269 |                     let state = state.lock().unwrap();
        |                                 ^^^^^-------
        |                                 |
        |                                 borrowed value does not live long enough
        |                                 argument requires that `state` is borrowed for `'static`
        |
    269 |                     let state = state.lock().unwrap();
        |                                 ^^^^^-------
        |                                 |
        |                                 borrowed value does not live long enough
        |                                 argument requires that `state` is borrowed for `'static`
    ...
    311 | }
        | - `state` dropped here while still borrowed

    error[E0597]: `state` does not live long enough
      --> src\proto\server\conn.rs:272:27
        |
    272 |                     match state.rpc_actions.get(&rpc.entry_id) {
        |                           ^^^^^-------------------------------
        |                           borrowed value does not live long enough
        |                           argument requires that `state` is borrowed for `'static`
    ...
    295 |                 }
        |                 - `state` dropped here while still borrowed

    error: aborting due to 2 previous errors

    For more information about this error, try `rustc --explain E0597`.
    error: could not compile `nt`.

    To learn more, run the command again with --verbose.

I can't figure out how to solve the ownership issue here. Any ideas?

**Comment by Redrield** [comment]
Are the pinned boxes necessary? I don't know if that's the issue, code isn't pushed yet, but it might be a part of it

**Comment by jlmcmchl** [comment]
The pinned box fixed the ownership issue for the UnboundSender `client`, since state has ownership over it and sending it to a new thread would require having no existing ownership in the message receive thread. At least, that's my understanding.

I'll push code this evening with two commits - one fixing the other changes, and then a working commit so you can see what I'm seeing.

**Comment by Redrield** [comment]
Futures channels are mpsc, you can clone it to the thread rather than having to box it.

**Comment by jlmcmchl** [comment]
@Redrield Pushed + took out the pin on the channel. Can you take a look?

**Comment by jlmcmchl** [comment]
Got it - updated rpc_actions to HashMap<u16,Arc<RpcAction>> and it compiles.

File: src/nt.rs [hunk 6]
```diff
@@ -139,6 +149,18 @@ impl NetworkTables<Server> {
...
+
+    pub fn create_rpc(
+        &mut self,
+        data: EntryData,
+        callback: impl Fn(Vec<u8>) -> Vec<u8> + Send + Sync + RefUnwindSafe + 'static,
+    ) {
+        self.state.lock().unwrap().create_rpc(data, callback);
+    }
+
+    pub fn call_rpc(&self, id: u16, parameter: Vec<u8>) -> Vec<u8> {
```
Comment (Redrield) [inline src/nt.rs]: Servers cannot initiate RPC Execute messages according to the spec

**Review by Redrield — CHANGES_REQUESTED** [review]
Servers can't initiate RPC, I missed this last time but just get rid of that.

Also, make sure that everything compiles and functions with the websocket feature enabled. Maybe make an RPC websocket client to make sure that it plays nicely with the server.

File: src/nt.rs (hunk 6 above)
Comment (jlmcmchl) [reply src/nt.rs]: fixed

File: src/proto/server/conn.rs (hunk 5 above)
Comment (jlmcmchl) [reply src/proto/server/conn.rs]: fixed

File: src/proto/server.rs (hunk 4 above)
Comment (jlmcmchl) [reply src/proto/server.rs]: fixed

File: src/proto/client.rs (hunk 3 above)
Comment (jlmcmchl) [reply src/proto/client.rs]: fixed

File: src/proto/client/conn.rs (hunk 2 above)
Comment (jlmcmchl) [reply src/proto/client/conn.rs]: fixed

**Review by Redrield — APPROVED** [review]

**Comment by Redrield** [comment]
Thank you for your contribution!

**Pull request merged**